python/ray/air/execution/resources/placement_group.py

```diff
@@ -0,0 +1,213 @@
+            start = -1
+            bundles = [{}] + self.resource_request.bundles
+        else:
+            start = 0
+            bundles = self.resource_request.bundles
+
+        if len(objects) > len(bundles):
+            raise RuntimeError(
+                f"The number of objects to annotate ({len(objects)}) cannot "
+                f"exceed the number of available bundles ({len(bundles)})."
+            )
+
+        annotated = []
+        for i, (obj, bundle) in enumerate(zip(objects, bundles), start=start):
+            bundle = bundle.copy()
+            num_cpus = bundle.pop("CPU", 0)
+            num_gpus = bundle.pop("GPU", 0)
+            memory = bundle.pop("memory", 0.0)
+
+            annotated.append(
+                obj.options(
+                    scheduling_strategy=PlacementGroupSchedulingStrategy(
+                        placement_group=self.placement_group,
+                        # Max ensures that empty head bundles are correctly placed
+                        placement_group_bundle_index=max(0, i),
+                        placement_group_capture_child_tasks=True,
+                    ),
+                    num_cpus=num_cpus,
+                    num_gpus=num_gpus,
+                    memory=memory,
+                    resources=bundle,
+                )
+            )
+        return annotated
+
+
+@DeveloperAPI
+class PlacementGroupResourceManager(ResourceManager):
+    """Resource manager using placement groups as the resource backend.
+
+    Ray core does not emit events when resources are available. Instead, we
+    have to periodically request and update the scheduling state of the
+    placement groups.
+
+    Internally, the placement group lifecycle is like this:
+    - Resources are requested with ``request_resources()``
+    - A placement group is scheduled ("staged")
+    - A ``PlacementGroup.ready()`` is scheduled ("staging future")
+    - We update the scheduling state when we need to
+        (e.g. when ``has_resources_ready()`` is called)
+    - When staging futures resolved, a placement group is moved from "staging"
+        to "ready"
+    - When a resource request is canceled, we remove a placement group from "staging".
+        If there are not staged placement groups (because they are already "ready"),
+        we remove one from "ready" instead.
+    - When a resource is acquired, the pg is removed from "ready" and moved
+        to "acquired"
+    - When a resource is freed, the pg is removed from "acquired" and destroyed
+
+    Per default, placement group scheduling state is refreshed every time when
+    resource state is inquired, but not more often than once every ``update_interval``
+    seconds. Alternatively, staging futures can be retrieved (and awaited) with
+    ``get_resource_futures()`` and state update can be force with ``update_state()``.
+
+    Args:
+        update_interval: Minimum interval in seconds between updating scheduling
```

Comment:
nit: Thoughts on adding the unit directly in the name to make it more explicit? e.g. `update_interval_s`

Reply:
Good call, updated